lib/repos/remote/knowns.go

```diff
@@ -0,0 +1,116 @@
+package remote
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"path"
+	"sync"
+)
+
+var (
+	knownGit = []string{
+		"github.com",
+		"gitlab.com",
+		"bitbucket.com",
+	}
+	knownOCI = []string{}
+)
+
+const knownsFileName = "hof-knowns.json"
+
+func knownsFilePath() (string, error) {
+	d, err := os.UserCacheDir()
+	if err != nil {
+		return "", fmt.Errorf("user cache dir: %w", err)
+	}
+
+	return path.Join(d, knownsFileName), nil
+}
+
+func NewKnowns() (*Knowns, error) {
+	p, err := knownsFilePath()
+	if err != nil {
+		return nil, fmt.Errorf("knowns file path: %w", err)
+	}
+
+	f, err := os.Open(p)
+	if err != nil {
+		return nil, fmt.Errorf("os open %s: %w", p, err)
+	}
+
+	defer f.Close()
+
+	var kn Knowns
+	if err := json.NewDecoder(f).Decode(&kn.values); err != nil {
+		return nil, fmt.Errorf("json decode %s: %w", p, err)
+	}
+
+	return &kn, nil
+}
+
+type Knowns struct {
+	valuesMu sync.RWMutex
```

Comment:
we can embed the sync type here and call the functions directly on our type

```
type Knowns struct {
	sync.RWMutex
```

```
	kn.RLock()
	defer kn.RUnlock()
```

Reply:
I'd prefer to avoid this pattern. It makes it difficult to reason about who is locking what (since the mutex API becomes part of our API).

Reply:
sgtm, wasn't thinking about Knowns being used outside the package